internal/provider/resource_apply_module.go

```diff
@@ -0,0 +1,573 @@
+						},
+						"key": schema.StringAttribute{
+							MarkdownDescription: "Key or name of this variable.",
+							Description:         "Key or name of this variable.",
+							Required:            true,
+						},
+						"category": schema.StringAttribute{
+							MarkdownDescription: "Category of this variable, 'terraform' or 'environment'.",
+							Description:         "Category of this variable, 'terraform' or 'environment'.",
+							Required:            true,
+						},
+						"hcl": schema.BoolAttribute{
+							MarkdownDescription: "Whether this variable is HCL (vs. environment).",
+							Description:         "Whether this variable is HCL (vs. environment).",
+							Required:            true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// Configure lets the provider implement the ResourceWithConfigure interface.
+func (t *applyModuleResource) Configure(_ context.Context,
+	req resource.ConfigureRequest, _ *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	t.client = req.ProviderData.(*tharsis.Client)
+}
+
+func (t *applyModuleResource) Create(ctx context.Context,
+	req resource.CreateRequest, resp *resource.CreateResponse) {
+
+	// Retrieve values from apply module.
+	var applyModule ApplyModuleModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &applyModule)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Do plan and apply, no destroy.
+	var didRun doRunOutput
+	resp.Diagnostics.Append(t.doRun(ctx, &doRunInput{
+		model: &applyModule,
+	}, &didRun)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Update the plan with the computed ID.
+	applyModule.ID = types.StringValue(uuid.New().String())
+	applyModule.ModuleVersion = types.StringValue(didRun.moduleVersion)
+
+	// Set the response state to the fully-populated plan, whether or not there is an error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, applyModule)...)
+}
+
+func (t *applyModuleResource) Read(ctx context.Context,
+	req resource.ReadRequest, resp *resource.ReadResponse) {
+
+	// Get the current state.
+	var state ApplyModuleModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var appliedModuleInfo appliedModuleInfo
+	resp.Diagnostics.Append(t.getCurrentApplied(ctx, state, &appliedModuleInfo)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Update the state with the computed module source and version.
+	if appliedModuleInfo.moduleSource != nil {
+		state.ModuleSource = types.StringValue(*appliedModuleInfo.moduleSource)
+	} else {
+		state.ModuleSource = types.StringNull()
+	}
+	if appliedModuleInfo.moduleVersion != nil {
+		state.ModuleVersion = types.StringValue(*appliedModuleInfo.moduleVersion)
+	} else {
+		state.ModuleVersion = types.StringNull()
+	}
+
+	// TODO: Eventually, when the API and SDK support speculative plan runs with a module source,
+	// this should do a speculative plan run here to determine whether changes are needed.
```

Comment:
Were we unable to do the plan (non-speculative) and pass along the id so that we could do the apply in the apply stage?